### PR TITLE
Editor bracket matching, ctrl+s saving and esc cancelclosing, editor color themes

### PR DESCRIPTION
**Pull request by Healarious**
I have decided to remove double and triple click word/line selection and do it in a future update

# Current additions
* ctrl + s closing and esc cancelclosing
* automatic bracket mathcing
* editor color themes

ss

# issues
* ~~the following bug appears when changing themes, [pastebin link](https://pastebin.com/n09TWUcf)~~
* ~~Some issues with bracket matching when pasting code~~
* ~~editor themes not being saved across sessions~~

### CODE REVIEW

**Comment by Healarious** [comment]
I have decided to remove double and triple click word/line selection and do it in a future update

# Current additions
* ctrl + s closing and esc cancelclosing
* automatic bracket mathcing
* editor color themes
* did some minor changes to the weakaura templates for cooldowns that weren''t in their correct specializations (for priest)

**Comment by Healarious** [comment]
I fucked up with the git commands

**Comment by Healarious** [comment]
got a a good copy again

**Comment by Healarious** [comment]
Ready for testing ( more color themes might be added in the future, gooling some themes I find appealing right now)

**Comment by Healarious** [comment]
# Ready for testing

**Comment by Stanzilla** [comment]
I think I'm fine with everything IF we can make the autocompletion an option the user has to turn on. The ideal place would be the `Themes` button you added and the dropdown it options, in the same move the button should probably say `Settings` instead of `Themes` then. I really want to see how that feels.

**Comment by Healarious** [comment]
Okay, I will make those edits right now

**Comment by Healarious** [comment]
Requested edits have been made

**Comment by Stanzilla** [comment]
It's a little derpy now https://gfycat.com/UnlawfulWellmadeAddax

**Comment by Healarious** [comment]
oh whoopsy, lel, lemme fix that

**Comment by Healarious** [comment]
the lil derp should be gone now

**Comment by Stanzilla** [comment]
Still a bit derpy https://gfycat.com/GoodnaturedTiredAmericanavocet

File: WeakAurasOptions/OptionsFrames/TextEditor.lua [hunk 1]
```diff
@@ -77,22 +108,109 @@ local function ConstructTextEditor(frame)
...
+    -- themes options
+    for k, v in pairs(WeakAurasSaved["editor_themes"]["themes"]) do
+      local item = {
+        text = k,
+        isNotRadio = false,
+        checked = (WeakAurasSaved["editor_themes"]["selected"] == k and true) or false,
+        func = function()
+          is_settings_open = false
+          WeakAurasSaved["editor_themes"]["selected"] = k
+          -- Caused mentioned overflow bug, the codeeditor text now has to be changed for the given theme to appear
```
Comment (Stanzilla) [inline WeakAurasOptions/OptionsFrames/TextEditor.lua]: is this still true? Comment still needed/correct?

**Comment by Stanzilla** [comment]
You will need to rebase this onto master in order to drop the template changes from this PR.

File: WeakAurasOptions/OptionsFrames/TextEditor.lua [hunk 2]
```diff
@@ -77,22 +109,106 @@ local function ConstructTextEditor(frame)
...
+  settings_frame:SetWidth(100)
+  settings_frame:SetText("Settings")
+  settings_frame:EnableMouse(true)
+  settings_frame:RegisterForClicks("AnyDown")
+
+  local menu_frame = CreateFrame("Frame", "SettingsMenuFrame", settings_frame, "UIDropDownMenuTemplate")
+  menu_frame:SetPoint("CENTER", settings_frame, "Center")
+  menu_frame:Hide()
+
+  local function settings_menu()
```
Comment (InfusOnWoW) [inline WeakAurasOptions/OptionsFrames/TextEditor.lua]: this recreates the menu table each time. That seems unnecessary.

File: WeakAurasOptions/OptionsFrames/TextEditor.lua [hunk 3]
```diff
@@ -77,22 +109,106 @@ local function ConstructTextEditor(frame)
...
+    return menu
+  end
+
+  settings_frame:SetScript("OnClick", function(self, button, down)
+    if button == "LeftButton" then
+      if is_settings_open then
+        ToggleDropDownMenu(1, nil, menu_frame, settings_frame, 0, 0, settings_menu(), nil, 0)
+        is_settings_open = false
+      else
+        EasyMenu(settings_menu(), menu_frame, settings_frame, 0, 0, "MENU")
```
Comment (InfusOnWoW) [inline WeakAurasOptions/OptionsFrames/TextEditor.lua]: I think it should be possible to use ToggleDropDownMenu to open the menu. That would remove the need for the is_settings_open variable and simplify the code a bit.

File: WeakAurasOptions/OptionsFrames/TextEditor.lua [hunk 4]
```diff
@@ -17,44 +17,76 @@ local textEditor
...
-  ["~="] = relationColor,
-  -- logical operators
-  ["and"] = logicColor,
-  ["or"] = logicColor,
-  ["not"] = logicColor,
-  -- misc
-  [0] = "|r",
-}
+if not WeakAurasSaved["editor_themes"] then
+  WeakAurasSaved["editor_themes"] = {
```
Comment (InfusOnWoW) [inline WeakAurasOptions/OptionsFrames/TextEditor.lua]: What the reason for saving the themes in the settings?

File: WeakAurasOptions/OptionsFrames/TextEditor.lua [hunk 5]
```diff
@@ -77,22 +109,106 @@ local function ConstructTextEditor(frame)
...
 
   local close = CreateFrame("Button", nil, group.frame, "UIPanelButtonTemplate");
   close:SetScript("OnClick", function() group:Close() end);
   close:SetPoint("RIGHT", cancel, "LEFT", -10, 0)
+  close:SetFrameLevel(close:GetFrameLevel() + 1)
   close:SetHeight(20);
   close:SetWidth(100);
   close:SetText(L["Done"]);
 
+  IndentationLib.enable(editor.editBox, get_scheme(WeakAurasSaved["editor_themes"]["selected"]), 4)
```
Comment (InfusOnWoW) [inline WeakAurasOptions/OptionsFrames/TextEditor.lua]: I prefer WeakAurasSaved.editor_themes.selected.

File: WeakAurasOptions/OptionsFrames/TextEditor.lua (hunk 4 above)
Comment (Healarious) [inline WeakAurasOptions/OptionsFrames/TextEditor.lua]: I did that initially so maybe ppl could make their own themes in the future something which I think is unnecesarry now

File: WeakAurasOptions/OptionsFrames/TextEditor.lua (hunk 3 above)
Comment (Healarious) [inline WeakAurasOptions/OptionsFrames/TextEditor.lua]: Will do

**Comment by Healarious** [comment]
All the requested changes have been made

File: WeakAurasOptions/OptionsFrames/TextEditor.lua [hunk 6]
```diff
@@ -143,46 +136,41 @@ local function ConstructTextEditor(frame)
...
-      table.insert(menu, item)
+    }
+    table.insert(menu, item)
     end
     -- bracket matching option
     table.insert(menu, {
       text = "Bracket Matching",
       isNotRadio = true,
-      checked = WeakAurasSaved["bracket_matching"],
+      checked = WeakAurasSaved.editor_bracket_mathcing,
```
Comment (Stanzilla) [inline WeakAurasOptions/OptionsFrames/TextEditor.lua]: typo

File: WeakAurasOptions/OptionsFrames/TextEditor.lua [hunk 7]
```diff
@@ -143,46 +136,41 @@ local function ConstructTextEditor(frame)
...
-      else
-        EasyMenu(settings_menu(), menu_frame, settings_frame, 0, 0, "MENU")
-        is_settings_open = true
-      end
+        ToggleDropDownMenu(1, nil, menu_frame, settings_frame, 0, 0, settings_menu(), nil, 27)
     end
   end)
 
-  -- bracket matching, saving (ctrl + s) and closing (esc)
+  -- bracket mathcing, saving (ctrl + s) and closing (esc)
```
Comment (Stanzilla) [inline WeakAurasOptions/OptionsFrames/TextEditor.lua]: typo

File: WeakAurasOptions/OptionsFrames/TextEditor.lua [hunk 8]
```diff
@@ -195,7 +183,7 @@ local function ConstructTextEditor(frame)
   end)
 
   editor.editBox:HookScript("OnChar", function(_, char)
-    if not IsControlKeyDown() and WeakAurasSaved["bracket_matching"] then
+    if not IsControlKeyDown() and WeakAurasSaved.editor_bracket_mathcing then
```
Comment (Stanzilla) [inline WeakAurasOptions/OptionsFrames/TextEditor.lua]: typo

File: WeakAurasOptions/OptionsFrames/TextEditor.lua [hunk 9]
```diff
@@ -143,46 +136,41 @@ local function ConstructTextEditor(frame)
...
     -- bracket matching option
     table.insert(menu, {
       text = "Bracket Matching",
       isNotRadio = true,
-      checked = WeakAurasSaved["bracket_matching"],
+      checked = WeakAurasSaved.editor_bracket_mathcing,
       func = function()
-        is_settings_open = false
-        WeakAurasSaved["bracket_matching"] = not WeakAurasSaved["bracket_matching"]
+        WeakAurasSaved.editor_bracket_mathcing = not WeakAurasSaved.editor_bracket_mathcing
```
Comment (Stanzilla) [inline WeakAurasOptions/OptionsFrames/TextEditor.lua]: typo

**Pull request merged**